Revit_Core_Engine/Convert/Revit/FromRevit/ParameterDefinition.cs

```diff
@@ -0,0 +1,101 @@
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using Autodesk.Revit.DB;
+using BH.Engine.Adapters.Revit;
+using BH.oM.Adapters.Revit.Parameters;
+using BH.oM.Adapters.Revit.Settings;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+using BH.oM.Adapters.Revit.Elements;
+
+namespace BH.Revit.Engine.Core
+{
+	public static partial class Convert
+	{
+		/***************************************************/
+		/****               Public Methods              ****/
+		/***************************************************/
+
+
+		[Description("Converts a ParameterElement to BH.oM.Adapters.Revit.Parameters.ParameterDefinition.")]
+		[Input("parameter", "Revit Parameter to be converted.")]
+		[Input("parameterLinks", "A collection of names of RevitParameters and sets of their correspondent Revit parameter names to be used on name mapping.")]
+		[Input("onlyLinked", "If true, there needs to be a valid, relevant parameter link in parameterLinks in order for convert to succeed.")]
+		[Output("parameterDefinition", "BH.oM.Adapters.Revit.Parameters.ParameterDefinition resulting from converting the input Revit Parameter.")]
+		public static ParameterDefinition ParameterElementFromRevit(this ParameterElement parameter, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
+		{
+			settings = settings.DefaultIfNull();
+			ParameterDefinition paramDefinition = refObjects.GetValue<ParameterDefinition>(parameter.Id);
+			if (paramDefinition != null)
+				return paramDefinition;
+
+			Definition def = parameter.GetDefinition();
+			Document document = parameter.Document;
+			Binding binding = document.ParameterBindings.get_Item(def);
+			ElementBinding elementBinding = binding as ElementBinding;
+
+			string name = parameter.Name;
+
+			string parameterType = def.GetDataType().TypeId;
+
+			ForgeTypeId groupTypeId = def.GetGroupTypeId();
+			string parameterGroup = LabelUtils.GetLabelForGroup(groupTypeId);
+
+			bool instance = elementBinding is InstanceBinding;
+
+			List<string> categories = new List<string>();
+			if (elementBinding?.Categories != null)
+			{
+				categories = elementBinding.Categories.Cast<Category>().Select(c => c.Name).ToList();
+			}
+
+			paramDefinition = new ParameterDefinition
+			{
+				Name = name,
+				ParameterType = parameterType,
+				ParameterGroup = parameterGroup,
+				Instance = instance,
+				Categories = categories,
+			};
+
+			if (parameter is SharedParameterElement sharedParameterElement)
+			{
+				paramDefinition.Shared = true;
+				paramDefinition.CustomData.Add("GUID", sharedParameterElement.GuidValue.ToString());
```

Comment:
As raised on previous occasions, I am a bit concerned about using `CustomData` as a platform to store information by design. I would rather consider one of the two instead:
- do not store this information at all
- adding an extra property to the `ParameterDefinition` class, which would be set on `Pull`, while on `Push` we'd need to add a warning in `ToRevitParameterElement` method in case the value is not empty